Implement schema matching inside Dispatcher

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35058 C++ smoketests for the API.
* #34668 Python-driven testing for the new operator registration API
* #34620 Minor OperatorName cleanup
* #34618 Improved error checking
* **#34601 Implement schema matching inside Dispatcher**
* #34456 Finish implementing schema-less impl, switch all internal registrations to it
* #34455 Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)
* #34221 Relax internal invariants of Dispatcher in preparation for schema-less registration

- Store the inferred function schema inside OperatorEntry so we
  can check when we find out the "real" schema
- Add fallback, CppFunction::makeFallthrough and
  CppFunction::makeFromBoxedFunction to public API, so we can
  stop calling internal registerImpl directly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396129](https://our.internmc.facebook.com/intern/diff/D20396129)

ezyang committed:
Implement schema matching inside Dispatcher

- Store the inferred function schema inside OperatorEntry so we
  can check when we find out the "real" schema
- Add fallback, CppFunction::makeFallthrough and
  CppFunction::makeFromBoxedFunction to public API, so we can
  stop calling internal registerImpl directly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Implement schema matching inside Dispatcher"


- Store the inferred function schema inside OperatorEntry so we
  can check when we find out the "real" schema
- Add fallback, CppFunction::makeFallthrough and
  CppFunction::makeFromBoxedFunction to public API, so we can
  stop calling internal registerImpl directly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Implement schema matching inside Dispatcher"


WARNING: This PR is KNOWN TO FAIL TESTS. #34618 is intended to the PR that passes all tests, and these will be squashed after review is finished.

- Store the inferred function schema inside OperatorEntry so we
  can check when we find out the "real" schema
- Add fallback, CppFunction::makeFallthrough and
  CppFunction::makeFromBoxedFunction to public API, so we can
  stop calling internal registerImpl directly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396129](https://our.internmc.facebook.com/intern/diff/D20396129)

[ghstack-poisoned]

ezyang committed:
Update on "Implement schema matching inside Dispatcher"


WARNING: This PR is KNOWN TO FAIL TESTS. #34618 is intended to the PR that passes all tests, and these will be squashed after review is finished.

- Store the inferred function schema inside OperatorEntry so we
  can check when we find out the "real" schema
- Add fallback, CppFunction::makeFallthrough and
  CppFunction::makeFromBoxedFunction to public API, so we can
  stop calling internal registerImpl directly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396129](https://our.internmc.facebook.com/intern/diff/D20396129)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/op_registration/op_registration.h:
```diff
@@ -662,6 +662,23 @@ class CAFFE2_API CppFunction final {
   }
 
+  // TODO: more user friendly API
+  static CppFunction makeFallthrough() {
```
Also: if fallbacks are lower precedence than op-specific kernels, why are fallthroughs needed? Maybe just to avoid empty dispatch table entries?

ezyang replied:
Obsoleted by comment above

bhosmer reviewed (approved):
LGTM - bunch of questions and a couple of nits inline, maybe one possibly substantive suggestion around error checking alias analysis modes.

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -47,10 +47,34 @@ void OperatorEntry::prepareForDeregistration() {
 
+namespace {
+  void checkSchema(const OperatorName& name, const FunctionSchema& expected, const FunctionSchema& actual) {
+    if (!actual.isDefaultAliasAnalysisKind()) {
```
Later in the patchset the check is simplified to its final form:

```
  void checkSchema(const OperatorName& name, const FunctionSchema& from_def, const FunctionSchema& inferred) {
    c10::optional<std::string> schema_difference = findSchemaDifferences(from_def, inferred);
    if (schema_difference.has_value()) {
      TORCH_CHECK(false,
        "In registration for ", toString(name), ": expected schema of operator to be \"", toString(from_def), "\", ",
        "but got inferred schema \"", toString(inferred), "\". ",
        *schema_difference);
    }
  }
```

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -47,10 +47,34 @@ void OperatorEntry::prepareForDeregistration() {
   TORCH_INTERNAL_ASSERT(!schema_.has_value());
   schema_ = std::move(schema);
   dispatchTable_.registerSchema(*schema_);
+  for (auto i = kernels_.begin(); i != kernels_.end(); ++i) {
```
I think I agree that we should be able to eliminate the default alias analysis kind, but I haven't done it in this patch set yet. I think we have to wait until I make "register schema with alias analysis kind; register schema without alias analysis kind" a hard error. I've held off on doing this as it basically implies that I have to fix every use site to use the new API.

bhosmer replied:
Sure, makes sense

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -59,21 +83,25 @@ void OperatorEntry::deregisterSchema() {
 }
 
-std::list<KernelFunction>::iterator OperatorEntry::registerKernel(c10::optional<DispatchKey> dispatch_key, KernelFunction kernel) {
+std::list<OperatorEntry::ListEntry>::iterator OperatorEntry::registerKernel(c10::optional<DispatchKey> dispatch_key, KernelFunction kernel, std::unique_ptr<FunctionSchema> inferred_function_schema) {
```
They were necessary in some cases and so I added them everywhere. They are required here, specifically. https://godbolt.org/z/HhzMDe

bhosmer replied:
Oh that's right, forgot it was a nested struct and we were outside the class scope. Thanks!

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -138,12 +138,12 @@ void Dispatcher::deregisterDef_(const OperatorHandle& op, const OperatorName& op
   auto op = findOrRegisterName_(op_name);
 
-  auto kernel_handle = op.operatorIterator_->op.registerKernel(dispatch_key, std::move(kernel));
+  auto kernel_handle = op.operatorIterator_->op.registerKernel(dispatch_key, std::move(kernel), std::move(inferred_function_schema));
```
I'm happy to rename this, but I don't think `kernel_handle` naming is a big deal, since I think of the schema as semantically being part of the kernel.

bhosmer replied:
Agree it's not a big deal - I think I navigate by these sorts of names more than most while I'm learning code, otherwise I doubt I'd even have noticed it :)

ezyang replied:
This is now `handle`

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -59,21 +83,25 @@ void OperatorEntry::deregisterSchema() {
-  k.push_front(std::move(kernel));
-  std::list<KernelFunction>::iterator inserted = k.begin();
+  k.emplace_front(std::move(kernel), std::move(inferred_function_schema));
+  std::list<OperatorEntry::ListEntry>::iterator inserted = k.begin();
```
...?

bhosmer replied:
Sorry, meant to cancel 😬

ezyang commented on aten/src/ATen/core/op_registration/op_registration.cpp:
```diff
@@ -189,14 +160,20 @@ Module&& Module::def(const char* name_or_schema_str, CppFunction&& f) && {
+  return std::move(*this);
+}
+
+Module&& Module::fallback(CppFunction&& f) && {
```
I didn't read your text carefully enough. I updated the linked discussion.

ezyang committed:
Update on "Implement schema matching inside Dispatcher"


WARNING: This PR is KNOWN TO FAIL TESTS. #34618 is intended to the PR that passes all tests, and these will be squashed after review is finished.

- Store the inferred function schema inside OperatorEntry so we
  can check when we find out the "real" schema
- Add fallback, CppFunction::makeFallthrough and
  CppFunction::makeFromBoxedFunction to public API, so we can
  stop calling internal registerImpl directly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396129](https://our.internmc.facebook.com/intern/diff/D20396129)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -47,10 +47,34 @@ void OperatorEntry::prepareForDeregistration() {
   dispatchTable_.registerSchema(*schema_);
+  for (auto i = kernels_.begin(); i != kernels_.end(); ++i) {
+    for (auto j = i->second.begin(); j != i->second.end(); ++j) {
+      if (j->inferred_function_schema) {
```
nit: I don't like implicit to-bool conversions, can we do `.has_value()`?

smessmer commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -16,6 +16,15 @@ namespace impl {
 // and its dispatch table. This is not part of the public API.
 class OperatorEntry final {
 public:
+  struct KernelSchemaPair {
```
nit: final

ezyang committed:
Update on "Implement schema matching inside Dispatcher"


- Store the inferred function schema inside OperatorEntry so we
  can check when we find out the "real" schema
- Add fallback, CppFunction::makeFallthrough and
  CppFunction::makeFromBoxedFunction to public API, so we can
  stop calling internal registerImpl directly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396129](https://our.internmc.facebook.com/intern/diff/D20396129)

[ghstack-poisoned]

ezyang committed:
Update on "Implement schema matching inside Dispatcher"


- Store the inferred function schema inside OperatorEntry so we
  can check when we find out the "real" schema
- Add fallback, CppFunction::makeFallthrough and
  CppFunction::makeFromBoxedFunction to public API, so we can
  stop calling internal registerImpl directly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396129](https://our.internmc.facebook.com/intern/diff/D20396129)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -138,12 +138,12 @@ void Dispatcher::deregisterDef_(const OperatorHandle& op, const OperatorName& op
 }
 
-RegistrationHandleRAII Dispatcher::registerImpl(OperatorName op_name, c10::optional<DispatchKey> dispatch_key, KernelFunction kernel) {
+RegistrationHandleRAII Dispatcher::registerImpl(OperatorName op_name, c10::optional<DispatchKey> dispatch_key, KernelFunction kernel, std::unique_ptr<FunctionSchema> inferred_function_schema) {
```
My editor line wraps long lines so things like function signatures look readable, that's my guilty plea XD

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.h:
```diff
@@ -16,6 +16,15 @@ namespace impl {
 // and its dispatch table. This is not part of the public API.
 class OperatorEntry final {
 public:
+  struct ListEntry {
```
It's now KernelEntry

ezyang commented on aten/src/ATen/core/dispatch/OperatorEntry.cpp:
```diff
@@ -47,10 +47,34 @@ void OperatorEntry::prepareForDeregistration() {
   schema_ = std::move(schema);
   dispatchTable_.registerSchema(*schema_);
+  for (auto i = kernels_.begin(); i != kernels_.end(); ++i) {
+    for (auto j = i->second.begin(); j != i->second.end(); ++j) {
```
https://github.com/pytorch/pytorch/issues/35041 but you are always going to have to keep a list to be commutative.

ezyang commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -138,24 +138,29 @@ void Dispatcher::deregisterDef_(const OperatorHandle& op, const OperatorName& op
   auto op = findOrRegisterName_(op_name);
 
-  auto kernel_handle = op.operatorIterator_->op.registerKernel(dispatch_key, std::move(kernel));
+  auto handle = op.operatorIterator_->op.registerKernel(dispatch_key, std::move(kernel), std::move(inferred_function_schema));
```
I really don't care, @bhosmer, fight it out with @smessmer XD

bhosmer replied:
_makes the hey man not a big deal gesture_ 👐

smessmer replied:
same for me, not a strong opinion ;)

ezyang commented:
All CRs in this PR were either responded to or applied locally to my patchset (not pushed yet)

ezyang committed:
Update on "Implement schema matching inside Dispatcher"


- Store the inferred function schema inside OperatorEntry so we
  can check when we find out the "real" schema
- Add fallback, CppFunction::makeFallthrough and
  CppFunction::makeFromBoxedFunction to public API, so we can
  stop calling internal registerImpl directly

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396129](https://our.internmc.facebook.com/intern/diff/D20396129)

[ghstack-poisoned]
